### PR TITLE
Add build template function

### LINKED ISSUE #6: Reverse token
Hi, what do you thinking about implement reverse tokenizer - from object to string with all params.

PS. You did really good job with this package.

### PR DESCRIPTION
**Pull request by kajetan-nobel**
It allows to build template with nested items based on given token.

Resolves #6

### CODE REVIEW

**Comment by mblarsen** [comment]
Hi @routegroup something I would really like to see was the possibility have it build the params again. Rarely do I turn a string to a toke and just back again without having changed the params; ignoring for a moment the children. 

Something like the reverse of `initParams` using the equivalent of castValue https://github.com/mblarsen/shortcode-tokenizer/blob/master/src/shortcode-tokenizer.js#L67 (eg. renderValue).

It could be:

```
buildTemplate(token: Token, params?: Object)
```

Usage

```
// should be real Token instance
const token = {
  ...
  name: 'product-list',
  params: {
     id: 42,
     title: "New Releases"
  },
  ...
}

buildTemplate(token) // [product-list id=42 title="New Releases" /]
buildTemplate(token, {}) // [product-list /]
token.params.title = "Just In"
buildTemplate(token) // [product-list id=42 title="Just In" /]
buildTemplate(token, {...token.params, {title: "Wow wow wow"}) // [product-list id=42 title="Wow wow wow" /]
```

Alternatively:

```
buildTemplate(token: Token, params?: Object | string)
```

Where params can be a pre-build string, but if object it behaves like above.

Thanks a lot for your contributions. I'll be looking forward to using this.

**Comment by routegroup** [comment]
@mblarsen but what about children? Maybe two methods about deep overwriting?

**Comment by kajetan-nobel** [comment]
Hi @mblarsen. Thank you for your response. I change a little bit my structure and refactor code to keep package code style.

I added:
- `token.toString([object|string|null params=null])` - you don't have to call `tokenizer.buildTemplate()` if you want to return string. Only one awareness - if it's open type it'll return `{slot}` between brackets - something like `[code]{slot}[/code]`
- `tokenizer.buildTemplate([Token], [object|string|null params=null], [bool deep = false])` -  allows right now to change by your proposition. The `deep` param will affect all nested children - so you can change only root or whenever you like with all nested params.

**Comment by kajetan-nobel** [comment]
I added also missed coverage tests and ignore coverage for depracted functions or really hard to get (like unknown token type).

**Comment by mblarsen** [comment]
@routegroup thanks for all your work. I would like to check out the code and I have a few changes. I'll do this tomorrow.

**Comment by mblarsen** [comment]
Can you in the meantime give me a practical use case for:

> The deep param will affect all nested children - so you can change only root or whenever you like with all nested params.

**Comment by kajetan-nobel** [comment]
@mblarsen When i thought about it was for rare use case. 

Example:

**CSS**
```css
.pa-3 {
  padding: 1rem;
}

```
**Input:**

```html
[row]
  [cell]Lorem ipsum[/cell]
  [cell]Lorem ipsum[/cell]
  [cell]Lorem ipsum[/cell]
  [cell]Lorem ipsum[/cell]
[/row]
```

**JS:**
```js
buildTemplate(input, { class: 'pa-3' }, true);
```

**Result:**
```html
[row class="pa-3"]
  [cell class="pa-3"]Lorem ipsum[/cell]
  [cell class="pa-3"]Lorem ipsum[/cell]
  [cell class="pa-3"]Lorem ipsum[/cell]
  [cell class="pa-3"]Lorem ipsum[/cell]
[/row]
```

File: src/shortcode-tokenizer.js [hunk 1]
```diff
@@ -326,6 +377,38 @@ export default class ShortcodeTokenizer {
...
+   * @throws {Error} Unexpected token type.
+   * @returns {string}
+   */
+  buildTemplate(token, params = null, deep = false) {
+    if (!token) {
+      return ''
+    }
+
+    if (!(token instanceof Token)) {
+      throw new Error('Expected Token instance.')
```
Comment (routegroup) [inline src/shortcode-tokenizer.js]: It's not necessary feature so feel free to remove that -

**Comment by kajetan-nobel** [comment]
@mblarsen it would also good idea to check if param is an array and if yes then join with space.

**Example**
```js
token.params.size = ['xs12', 'sm6', 'md4', 'lg3'];

// result
[item size="xs12 sm6 md4 lg3"]...[/item]
```

**Comment by mblarsen** [comment]
Sorry I don't find the deepParam useful enough to add it to the API. You can always map over the children prior to calling buildTemplate.

In addition I had the buildParam render the value correctly. I realized that there was a problem with master as well the other way around. Casting, this has been fixed and rebased. Be sure to:

```
git refresh --all ; git reset --hard origin/build-template
```

**Comment by mblarsen** [comment]
I'm published a new version 0.6.1 on NPM thanks for your contribution 🎉

**Pull request merged**